File: leetcode/medium/713-subarray-product-less-than-k.py
from typing import List


class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
        subarray_count = 0
        left = 0
        total = 1
        for right in range(len(nums)):
            total *= nums[right]
            if total >= k:
                while left <= right and total >= k:
                    total /= nums[left]
                    left += 1
            subarray_count += right - left + 1

        return subarray_count

File: leetcode/medium/713-subarray-product-less-than-k.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var numSubarrayProductLessThanK = function(nums, k) {
    let subarrayCount = 0;
    let left = 0;
    let total = 1;
    for(let right = 0; right < nums.length; right++) {
        total *= nums[right];
        if(total >= k) {
            while(left <= right && total >= k) {
                total /= nums[left];
                left++;
            }
        }
        subarrayCount += right - left + 1;
    }
    return subarrayCount;
};
